link-to-code/link-to-code
# Create coding exercise template API

## Description

This PR adds the coding exercise template API.
This API receives the details of a coding exercise and after performing all the validation and data consistency checks, creates a new coding exercise template and returns a permalink, to be used as the link of the web app.

A coding exercise template will then be used to create many interview rooms when clicking on the permalink.
An interview room will have a copy of the coding exercise details, to avoid making changes to the original data in the template (effectively acting like a fork).

## How to test

1) Locally create an `.env` file in the `apps/functions` folder and add an `ADMIN_SECRET=` entry with a plain password
2) Then run the `rush install && rush build` command in the root
3) In the `apps/functions` folder, run the `rushx emulator:start`
4) In a new terminal, create a base64 token for the secret by running `echo -n [YOUR ADMIN_SECRET HERE] | openssl base64`
5) Make a request running

```
curl --request POST \
  --url http://127.0.0.1:5001/link-to-code/us-central1/createCodingExerciseTemplate \
  --header 'Authorization: Bearer [REPLACE WITH YOUR BASE64 ADMIN_SECRET TOKEN HERE]' \
  --header 'Content-Type: application/json' \
  --data '{
	"codingExercise": {
		"name": "test4",
		"files": [{
			"filename": "index.js",
			"readonly": false,
			"language": "js",
			"content": "coosle.log('\''hello!'\'')"
		}],
		"entry": "index.js"
	}
}'
```

You should see in the emulator http://127.0.0.1:4000/firestore/data:
![image](https://user-images.githubusercontent.com/536973/198154516-28555c37-3988-4c84-92a5-bd921d986cc1.png)

And the response body should be:
![image](https://user-images.githubusercontent.com/536973/198154560-b39325ca-cc51-41c7-888f-4243658b2623.png)
